commit e116b703f0b9b551cb413c8f2cfe231a9ce73f42
Author: Michael Paquier <michael@paquier.xyz>
Date:   Tue Dec 17 09:23:49 2024 +0900

    pg_combinebackup: Fix PITR comparison test in 002_compare_backups
    
    The test was creating both the dumps to compare from the same database
    on the same node, so it would never detect any mismatches when comparing
    the logical dumps of the two servers.
    
    Fixing this issue has revealed that there is a difference in the dumps:
    the tablespaces paths are different.  This commit uses compare_text()
    with a custom comparison function to erase the difference (slightly
    tweaked to be able to work with WIN32 and non-WIN32 paths).  This way,
    the non-relevant parts of the tablespace path are ignored from the check
    with the basic structure of the query string still compared.
    
    Author: Dagfinn Ilmari Mannsåker
    Discussion: https://postgr.es/m/87h67653ns.fsf@wibble.ilmari.org
    Backpatch-through: 17

diff --git a/src/bin/pg_combinebackup/t/002_compare_backups.pl b/src/bin/pg_combinebackup/t/002_compare_backups.pl
index 63a0255de15..cfdd25471cb 100644
--- a/src/bin/pg_combinebackup/t/002_compare_backups.pl
+++ b/src/bin/pg_combinebackup/t/002_compare_backups.pl
@@ -1,10 +1,10 @@
 # Copyright (c) 2021-2024, PostgreSQL Global Development Group
 
 use strict;
 use warnings FATAL => 'all';
-use File::Compare;
+use File::Compare qw(compare_text);
 use PostgreSQL::Test::Cluster;
 use PostgreSQL::Test::Utils;
 use Test::More;
 
 my $tempdir = PostgreSQL::Test::Utils::tempdir_short();
@@ -173,21 +173,27 @@ $pitr1->command_ok(
 		$dump1, '--no-sync',
 		'--no-unlogged-table-data', '-d',
 		$pitr1->connstr('postgres'),
 	],
 	'dump from PITR 1');
-$pitr1->command_ok(
+$pitr2->command_ok(
 	[
 		'pg_dumpall', '-f',
 		$dump2, '--no-sync',
 		'--no-unlogged-table-data', '-d',
-		$pitr1->connstr('postgres'),
+		$pitr2->connstr('postgres'),
 	],
 	'dump from PITR 2');
 
-# Compare the two dumps, there should be no differences.
-my $compare_res = compare($dump1, $dump2);
+# Compare the two dumps, there should be no differences other than
+# the tablespace paths.
+my $compare_res = compare_text(
+	$dump1, $dump2,
+	sub {
+		s{create tablespace .* location .*\btspitr\K[12]}{N}i for @_;
+		return $_[0] ne $_[1];
+	});
 note($dump1);
 note($dump2);
 is($compare_res, 0, "dumps are identical");
 
 # Provide more context if the dumps do not match.
